### PR TITLE
Build Budget App

### PR DESCRIPTION
**Pull request by didierganthier**
## Checklist
- [x] I have read this repository's contribution guidelines.
- [x] My pull request has a descriptive title (not a vague title like Update index.MD)
- [x] My pull request targets the main branch of this [project](https://github.com/didierganthier/budget_app).
- [x] I added relevant documentation for all the files I changed

## Summary
This PR builds the budget app, a Mobile web application where you can manage your budget: you have a list of transactions associated with a category so that you can see how much money you spent and on what. Built with Ruby on Rails, TailwindCSS, and PostgreSQL. Using Devise for authentication and RSpec/Capybara for testing.

 ## What changes have been made?
- [x] Use ruby gems as software packages system.
- [x] Install Ruby on Rails framework.
- [x] Understand Rails RESTful design and router.
- [x] Use controllers to handle requests and render empty views.
- [x] Use params from browser request in a safe way.
- [x] Use preprocessed html file with embedded Ruby code.
- [x] Use layouts and templates for shared content.
- [x] Use database migration files to maintain database schema.
- [x] Use validations for models.
- [x] Secure app from n+1 problems.
- [x] Understand what ORM is.
- [x] Write SQL queries with ActiveRecord.
- [x] Set up associations between models.
- [x] Build a webapp that requires the user to log in.
- [x] Use devise gem for authentication.
- [x] Limit access to webapp resources based on authorization rules.

## Demo Video
https://www.loom.com/share/21547e1e518f43d599c4d569ee58dbcc

## Live Link
https://budget-app-rails.onrender.com

## Rspec tests
- User spec
Screenshot 2023-02-15 at 11 56 53
- Expense spec
Screenshot 2023-02-15 at 11 57 42
- Category spec
Screenshot 2023-02-15 at 11 58 26

## Capybara tests
Screenshot 2023-02-15 at 11 54 59

## Author
- @didierganthier

### CODE REVIEW

**Comment by KDlamini** [comment]
## Status: Invalid 🔴 

### Hi @didierganthier 🙋‍♀️,  

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request
 
You have submitted a project, that is not following the proper GitHub flow. Please merge this PR and submit a new PR merging the `dev` branch to the `main`. You should always submit a PR with the `dev` branch for your Capstone projects.

- Kindly add the relevant project name to your README

image

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏
 
Cheers 🥂  and Happy coding!!! 👯 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @KDlamini  in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_Invalid Code Review Request **does not count** into the code reviews limit._

File: README.md [hunk 1]
```diff
@@ -241,48 +159,40 @@ Feel free to check the [issues page](../../issues/).
...
 <p align="right">(<a href="#readme-top">back to top</a>)</p>
 
-<!-- ACKNOWLEDGEMENTS -->
-
 ## 🙏 Acknowledgments <a name="acknowledgements"></a>
 
-> Give credit to everyone who inspired your codebase.
-
-I would like to thank...
+I would like to thank Microverse for the opportunity to learn and grow as a developer.
```
Comment (AnselemOdims) [inline README.md]: - [ ] Please kindly include section that gives appropriate credit to [Gregoire Vella on Behance](https://www.behance.net/gregoirevella) the [author of the original design](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=), as required by the [Creative Commons license of the design](https://creativecommons.org/licenses/by-nc/4.0/).

File: app/views/categories/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+      <div class="flex flex-col items-center p-6 rounded-lg bg-white shadow-lg">
+        <%= link_to category.name, category_path(category), class: "text-xl font-medium mb-3" %>
+        <%= image_tag(category.icon, alt: "Icon pic", class: "w-12 h-12") %>
+        $<%= category.expenses.sum(:amount) %> spent
+      </div>
+    <% end %>
+    </div>
+  <% end %>
+  <%= link_to "Add new category", new_category_path, class: "bg-blue-400 text-white py-2 px-4 rounded-lg mt-6 hover:bg-blue-800" %>
+</div>
```
Comment (AnselemOdims) [inline app/views/categories/index.html.erb]: - [ ] Please kindly ensure your design matches the original design for the categories page.

### Your Design
image

### Original Design
image

File: app/views/categories/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+        <p class="text-lg font-medium mb-3"><%= expense.name %></p>
+        <p class="text-lg font-medium mb-3"><%= expense.amount %></p>
+        <p class="text-lg font-medium mb-3"><%= expense.created_at.strftime("%d %b %Y") %></p>
+      </div>
+    <% end %>
+  </div>
+<% end %>
+  <%= link_to "Add new expense", new_category_expense_path(@category), class: "bg-blue-400 text-white py-2 px-4 rounded-lg m-6" %>
+  <%= link_to "Back", categories_path, class: "text-blue-500 py-2 px-4 rounded-lg m-6" %>
+</div>
```
Comment (AnselemOdims) [inline app/views/categories/show.html.erb]: - [ ] As previously pointed out, please ensure your design matches the original design

### Your Design
image

### Original Design
image

**Review by AnselemOdims — CHANGES_REQUESTED** [review]
### Hi @didierganthier , 
You have been doing a tremendous job so far :ok_hand:
There are some issues that you still need to work on to go to the next project but you are almost there!

## To Highlight 🎉
PR has a title and summary. :heavy_check_mark:
No linter errors :heavy_check_mark:

## Required Changes ♻️
- [ ] Please also check inline comments under the review

## Optional suggestions
_Comments with the [OPTIONAL] prefix are not a must to implement however, It's strongly recommended to work on them as they can make your code better._

## Happy coding!👏👏👏

#### Please leave any questions or comments in the PR if you have any doubts or unclarity! You can tag me using @AnselemOdims for a quicker response

**_Note: Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: README.md (hunk 1 above)
Comment (didierganthier) [reply README.md]: Thanks for pointing that out 😉 , I will handle that.

File: app/views/categories/index.html.erb (hunk 2 above)
Comment (didierganthier) [reply app/views/categories/index.html.erb:18]: Thanks @AnselemOdims, i fixed it 😉 
Screenshot 2023-02-15 at 13 32 17

File: app/views/categories/show.html.erb (hunk 3 above)
Comment (didierganthier) [reply app/views/categories/show.html.erb]: This was handled 😉 
Screenshot 2023-02-15 at 13 43 18

**Review by VICTOR-CHUKWU — APPROVED** [review]
Hi @didierganthier ,
# STATUS: APPROVED! 🥇 💯 
done

In my opinion, your project is ready for the presentation!  There is nothing else to say other than... it's time to merge it :shipit: 


Good luck! 🎉 


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**Pull request merged**